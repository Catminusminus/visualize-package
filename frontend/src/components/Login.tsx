/**
 * Author of this code work, Yuki Mogi. c FFRI, Inc. 2019
 */

import * as React from 'react'
import Button from '@material-ui/core/Button'
import { clientId, redirectUrl } from '../utils/config'
import { oauthLoginUrl } from '@octokit/oauth-login-url'
import Tooltip from '@material-ui/core/Tooltip'

interface Props {
  state: string
}

const LoginButton: React.FC<Props> = ({ state }) => {
  const { url } = oauthLoginUrl({
    clientId,
    redirectUrl,
    scopes: ['repo'],
    state,
    log: {
      warn(message) {
        console.log(message, { level: 'warn' })
      }
    }
  })

  return (
    <div>
      <Tooltip title="Enable GitHub OAuth" aria-label="login">
        <Button variant="contained" href={url} disabled={!clientId}>
          Authorize App
        </Button>
      </Tooltip>
    </div>
  )
}

export default LoginButton
