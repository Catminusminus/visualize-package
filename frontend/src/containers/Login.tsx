/**
 * Author of this code work, Yuki Mogi. c FFRI, Inc. 2019
 */

import * as React from 'react'
import LoginButton from '../components/Login'
import LogoutButton from '../components/Logout'
import { clientId } from '../utils/config'
import { useSelector, useDispatch } from 'react-redux'
import { addError, State, setLoggedIn } from '../modules'
import { push } from 'connected-react-router'
import useSessionstorage from '@rooks/use-sessionstorage'
import useCookie from 'react-use-cookie'

const genState = () =>
  Array.from(window.crypto.getRandomValues(new Uint32Array(40)), num =>
    num.toString(16)
  ).join('') as any

const isLoggedInSelector = (state: State) => state.reducer.isLoggedIn

const Login = () => {
  const isLoggedIn = useSelector(isLoggedInSelector)
  const dispatch = useDispatch()
  const [state, setState] = useSessionstorage('state', genState())
  const [userID] = useCookie('user_id', '')
  React.useEffect(() => {
    if (!isLoggedIn && userID && clientId) {
      dispatch(
        setLoggedIn({
          isLoggedIn: true
        })
      )
    }
    // eslint-disable-next-line
  }, [dispatch, userID])
  const onClick = () => {
    setState(genState())
    dispatch({
      type: 'DELETE_TOKEN_SAGA',
      payload: {
        clientId
      }
    })
  }
  const query = window.location.search.substring(1)
  const requestToken = query.split('code=')
  if (requestToken.length !== 1 && !isLoggedIn) {
    const requestTokenWithoutState = requestToken[1].split('&state=')
    const code = requestTokenWithoutState[0]
    const returnedState = requestTokenWithoutState[1]
    dispatch(push('/'))
    if (returnedState !== state) {
      setState(genState())
      dispatch(
        addError({
          err: {
            id: 'GitHub OAuth',
            msg:
              'GitHub OAuth failed because the state parameter is incorrect. Close the session or restart the app.'
          }
        })
      )
    } else {
      dispatch({
        type: 'FETCH_TOKEN_SAGA',
        payload: {
          clientId,
          code
        }
      })
    }
  }
  return (
    <>
      {!isLoggedIn ? (
        <LoginButton state={state} />
      ) : (
        <LogoutButton onClick={onClick} />
      )}
    </>
  )
}

export default Login
