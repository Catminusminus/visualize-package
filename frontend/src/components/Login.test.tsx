/**
 * Author of this code work, Yuki Mogi. c FFRI, Inc. 2019
 */

import * as React from 'react'
import renderer from 'react-test-renderer'
import Login from './Login'

it('Snapshot test of Login', () => {
  const result = renderer.create(<Login state="abcdefg" />).toJSON()
  expect(result).toMatchSnapshot()
})
