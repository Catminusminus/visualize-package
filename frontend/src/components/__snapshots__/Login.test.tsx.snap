// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test of Login 1`] = `
<div>
  <a
    aria-describedby={null}
    aria-disabled={false}
    aria-label="login"
    className="MuiButtonBase-root MuiButton-root MuiButton-contained"
    href="https://github.com/login/oauth/authorize?allow_signup=true&client_id=aaaaaaaa&redirect_url=http://localhost:1234/&scope=repo&state=abcdefg"
    onBlur={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    title="Enable GitHub OAuth"
  >
    <span
      className="MuiButton-label"
    >
      Authorize App
    </span>
    <span
      className="MuiTouchRipple-root"
    />
  </a>
</div>
`;
